Add Digital Control chapter

martinjaeger opened the pull request:
This PR adds the content for above chapter.

Rendered version is here: https://learn.libre.solar/b/digital-control/development/digital_control.html

I started with some more general introductions (or recaps) about control and ended with an example implementation of a digital controller in C.

So far I didn't add any Bode plots because the hardware development and actual implementation is still ongoing at the moment. We can add some plots in the future after verification in actual hardware.

I hope the control experts don't find too many issues regarding nomenclature or over-simplifications in the text :)

rbtnx reviewed (commented):
Very nice summary of this quite complicated topic!
Here are a few comments/suggestions:

- Section analog vs digital control -> comparison
`Nonlinearities of analog components can be compensated.` - Maybe a bit confusing regarding the example with the op-amp, which is able to "compensate" the nonlinearities of the mosfet. Do you refer to the nonlinearity of the op-amp itself? If so, maybe add a short sentence to adress this issue the analog control section, eg. saturation

- Section Discrete time model
`The new matrices for the discretized model (suffix ddd) can be calculated by hand, but it's typically done with software tools like Matlab or Octave.` - Maybe add a short sentence and/or a link to the most common discretization method [Zero order hold](https://en.wikipedia.org/wiki/Zero-order_hold)

-  Section Digital controller
I think we should rather show an implementation on an actual PID controller showing the integral, proportional and derivative part and not using pre calculated variables A0, A1, A2. [Wikipedia](https://en.wikipedia.org/wiki/PID_controller) has a pseudo code example of an algorithm implementation, maybe something like that using pseudo-c?

martinjaeger commented:
@rbtnx thanks for your review!

> * Section analog vs digital control -> comparison
>   `Nonlinearities of analog components can be compensated.` - Maybe a bit confusing regarding the example with the op-amp, which is able to "compensate" the nonlinearities of the mosfet. Do you refer to the nonlinearity of the op-amp itself? If so, maybe add a short sentence to adress this issue the analog control section, eg. saturation

Good point. I think I was more thinking of any kind of nonlinear system behavior. For example if we have a nonlinear sensor reading, we can use a lookup table to convert it to something more useful for the control... how would you describe this? And is it actually true that this is a major advantage of digital control?

> * Section Discrete time model
>   `The new matrices for the discretized model (suffix ddd) can be calculated by hand, but it's typically done with software tools like Matlab or Octave.` - Maybe add a short sentence and/or a link to the most common discretization method [Zero order hold](https://en.wikipedia.org/wiki/Zero-order_hold)

Good idea, will add it.

> * Section Digital controller
>   I think we should rather show an implementation on an actual PID controller showing the integral, proportional and derivative part and not using pre calculated variables A0, A1, A2. [Wikipedia](https://en.wikipedia.org/wiki/PID_controller) has a pseudo code example of an algorithm implementation, maybe something like that using pseudo-c?

Yeah, maybe the additional step to make the calculation faster is not really necessary for the understanding. So should we just use C code with floating points and explain in the text that actual implemenations on smaller MCUs w/o floating point units require more advanced fixed-point math to make it fast?

rbtnx commented:
> Good point. I think I was more thinking of any kind of nonlinear system behavior. For example if we have a nonlinear sensor reading, we can use a lookup table to convert it to something more useful for the control... how would you describe this? And is it actually true that this is a major advantage of digital control?

Yeah, with a digital controller you could develop an algorithm that is capable of controlling a nonlinear plant (lookup tables are a very simple method, there are ways to actually compute a fitting input signal). To my understanding in the current sink example the op-amp acts a linear controller around the operational point - because there is no computation involved it's very fast but the efficiency is not very high. The digital controller on the other hand can regulate the system more efficiently (eg using a wider range of Vin) depending on the algorithm. I can check if the assumption about the op-amp acting as linearizing controller is true and can adjust the section appropriately if you want..

> > * Section Digital controller
> >   I think we should rather show an implementation on an actual PID controller showing the integral, proportional and derivative part and not using pre calculated variables A0, A1, A2. [Wikipedia](https://en.wikipedia.org/wiki/PID_controller) has a pseudo code example of an algorithm implementation, maybe something like that using pseudo-c?
> 
> Yeah, maybe the additional step to make the calculation faster is not really necessary for the understanding. So should we just use C code with floating points and explain in the text that actual implemenations on smaller MCUs w/o floating point units require more advanced fixed-point math to make it fast?

Yeah, I guess that's a good option :)

Gretel5X commented on docs/development/digital_control.md:
```diff
@@ -1,5 +1,310 @@
+
+<fig-caption src="development/digital-current-control.svg" caption="Digitally controlled current sink" num="3" />
+
+In this case, the output voltage is measured by an analog-to-digital converter (ADC), so that it can be used for calculations inside the microcontroller. The set-point could either be measured as a voltage by the ADC aswell or it can be defined by any other sort of input signal to the microcontroller, e.g. a communication interface.
```
.. by the ADC as well....
```suggestion
In this case, the output voltage is measured by an analog-to-digital converter (ADC), so that it can be used for calculations inside the microcontroller. The set-point could either be measured as a voltage by the ADC as well or it can be defined by any other sort of input signal to the microcontroller, e.g. a communication interface.
```

martinjaeger commented:
I have finally addressed all comments. Thanks for your reviews @rbtnx and @Gretel5X

For some reason it's not updating the rendered version behind above link. But I'm gonna merge anyway :)
